File: #0241.different-ways-to-add-parentheses.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string expression) {

        int length = expression.length();
        vector<int> result;
        char c;
        vector<int> left, right;
        for (int i = 0; i < length; i++) {
            c = expression[i];
            if (c != '+' && c != '-' && c != '*') {
                continue;
            }

            left = diffWaysToCompute(expression.substr(0, i));
            right = diffWaysToCompute(expression.substr(i + 1));
            for (auto l : left) {
                for (auto r : right) {
                    if (c == '+') {
                        result.push_back(l + r);
                    } else if (c == '-') {
                        result.push_back(l - r);
                    } else {
                        result.push_back(l * r);
                    }
                }
            }
        }

        if (!result.empty()) {
            return result;
        }

        int i = 0;
        for(auto e : expression){
            i = i * 10 + (e - '0');
        }
        result.push_back(i);
        return result;
    }
};

File: #0241.different-ways-to-add-parentheses.js
/**
 * @param {string} expression
 * @return {number[]}
 */
var diffWaysToCompute = function(expression) {
    function backtrack(nums, ops) {
        if (nums.length == 1) {
            return nums;
        }

        let result = [];
        for (let i = 0; i < ops.length; i++) {
            const left = backtrack(nums.slice(0, i + 1), ops.slice(0, i));
            const right = backtrack(nums.slice(i + 1), ops.slice(i + 1));
            for (let j of left) {
                for (let k of right) {
                    if (ops[i] == '*') {
                        result.push(j * k);
                    } else if (ops[i] == '+') {
                        result.push(j + k);
                    } else {
                        result.push(j - k);
                    }
                }
            }
        }
        return result;
    }

    const nums = expression.match(/\d+/g).map(Number);
    const ops = expression.match(/[^\d]/g);
    return backtrack(nums, ops);
};
